File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def deleteMiddle(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        
        if head.next == None:
            return None
        if head.next.next == None:
            head.next = None
            return head
        
        fast = head
        slow = head
        deleted = head
        
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            
        while deleted.next != slow:
            deleted = deleted.next
            

        deleted.next = slow.next
        return head
        
        
       

File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteMiddle(ListNode head) {
        
        if(head.next == null) return null;
        if(head.next.next == null){
            head.next = null;
            return head;
        } 
        
        ListNode fast = head, slow = head, deleted = head;
        while(fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }
        while(deleted.next != slow){
            deleted = deleted.next;
        }
        
        deleted.next = slow.next;
        return head;
        
        
        
    }
}


 